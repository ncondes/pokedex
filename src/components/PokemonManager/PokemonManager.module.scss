@use '@/scss/mixins' as *;

.tools {
   margin: 2rem auto;
   display: grid;
   grid-template-rows: repeat(3, 1fr);
   grid-template-columns: 1fr;

   @include breakpoint-up('md') {
      grid-template-rows: repeat(2, 1fr);
      grid-template-columns: repeat(2, 1fr);
      grid-gap: 0 1rem;
   }

   @include breakpoint-up('lg') {
      grid-template-rows: 1fr;
      grid-template-columns: 4fr 1fr 1fr;
      grid-gap: 1rem;
   }
}
