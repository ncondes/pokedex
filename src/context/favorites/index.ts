export { FavoritesContext } from './FavoritesContext'
export { FavoritesProvider } from './FavoritesProvider'
