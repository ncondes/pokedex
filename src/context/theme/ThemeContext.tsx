'use client'

import { createContext } from 'react'

interface Props {
   theme: 'light' | 'dark'
   toggleTheme: () => void
   useTheme: (className: string) => string
}

export const ThemeContext = createContext({} as Props)
