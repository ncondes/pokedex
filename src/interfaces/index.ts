export * from './data'
export * from './pokemon'
