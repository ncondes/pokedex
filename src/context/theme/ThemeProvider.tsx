'use client'

import { FC, PropsWithChildren, useEffect, useState } from 'react'
import { ThemeContext } from './ThemeContext'
import { getLocalTheme, setLocalTheme } from '@/utils'

export const ThemeProvider: FC<PropsWithChildren> = ({ children }) => {
   const [theme, setTheme] = useState<'light' | 'dark'>('light')

   const toggleTheme = () => {
      const t = theme === 'light' ? 'dark' : 'light'
      setTheme(t)
      setLocalTheme(t)
   }

   const useTheme = (className: string): string => `${theme}-${className}`

   useEffect(() => {
      const initialTheme = getLocalTheme() as 'light' | 'dark'
      setTheme(initialTheme)
   }, [])

   return (
      <ThemeContext.Provider value={{ theme, toggleTheme, useTheme }}>
         {children}
      </ThemeContext.Provider>
   )
}
