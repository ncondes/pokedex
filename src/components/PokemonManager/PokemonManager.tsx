'use client'

import { SearchBox } from '@/components/SearchBox'
import { Select } from '@/components/Select'
import { Pokemon } from '@/interfaces'
import { FC, useContext } from 'react'
import { useFilteredPokemons } from '@/hooks'
import { PokemonList } from '@/components/PokemonList'
import { ThemeContext } from '@/context/theme'
import { Introduction } from '../Introduction'
import styles from './PokemonManager.module.scss'

interface Props {
   data: Pokemon[]
}

const pokemonTypes = [
   'all',
   'bug',
   'dark',
   'dragon',
   'electric',
   'fairy',
   'fighting',
   'fire',
   'flying',
   'ghost',
   'grass',
   'ground',
   'ice',
   'normal',
   'poison',
   'psychic',
   'rock',
   'steel',
   'water',
]

const sortOptions = ['ascending', 'descending']

export const PokemonManager: FC<Props> = ({ data }) => {
   const { pokemons, handleSearch, handleFilter, handleSort } =
      useFilteredPokemons(data)

   const { useTheme } = useContext(ThemeContext)

   return (
      <>
         <Introduction />
         <div className={styles.tools}>
            <SearchBox callback={handleSearch} />
            <Select
               label="filter"
               icon="fa-filter"
               options={pokemonTypes}
               onChange={handleFilter}
            />
            <Select
               label="sort"
               icon="fa-sort"
               options={sortOptions}
               onChange={handleSort}
            />
         </div>
         <PokemonList pokemons={pokemons} />
      </>
   )
}
