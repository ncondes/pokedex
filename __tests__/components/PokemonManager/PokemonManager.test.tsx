import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'
import { PokemonManager } from '@/components/PokemonManager'
import { Pokemon, Stat } from '@/interfaces'
import { ThemeProvider } from '@/context/theme'

global.structuredClone = (val) => JSON.parse(JSON.stringify(val))

const pokemons = [
   {
      id: 1,
      name: 'Pikachu',
      types: [{ type: { name: 'electric' } }],
      sprites: {
         front_default:
            'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/25.png',
      },
      stats: [] as Stat[],
   },
   {
      id: 2,
      name: 'Charizard',
      types: [{ type: { name: 'fire' } }],
      sprites: {
         front_default:
            'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/6.png',
      },
      stats: [] as Stat[],
   },
   {
      id: 3,
      name: 'Bulbasaur',
      types: [{ type: { name: 'grass' } }],
      sprites: {
         front_default:
            'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png',
      },
      stats: [] as Stat[],
   },
] as Pokemon[]

describe('PokemonManager', () => {
   it('renders the component and handles search, filter, and sort actions', async () => {
      const { getByText, getByRole, getAllByRole, queryByText } = render(
         <ThemeProvider>
            <PokemonManager data={pokemons} />
         </ThemeProvider>
      )

      // initial rendering
      expect(getByText('Pikachu')).toBeInTheDocument()
      expect(getByText('Charizard')).toBeInTheDocument()
      expect(getByText('Bulbasaur')).toBeInTheDocument()

      const select = getAllByRole('combobox')[0]

      // filter action
      fireEvent.change(select, { target: { value: 'grass' } })

      const pikachu = queryByText('Pikachu')
      const charizard = queryByText('Charizard')
      const bulbasaur = queryByText('Bulbasaur')

      expect(pikachu).not.toBeInTheDocument()
      expect(charizard).not.toBeInTheDocument()
      expect(bulbasaur).toBeInTheDocument()
   })

   it('should match snapshot', () => {
      const { container } = render(
         <ThemeProvider>
            <PokemonManager data={pokemons} />
         </ThemeProvider>
      )

      expect(container).toMatchSnapshot()
   })
})
