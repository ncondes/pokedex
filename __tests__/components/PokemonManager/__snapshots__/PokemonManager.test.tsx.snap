// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonManager should match snapshot 1`] = `
<div>
  <p
    class="introduction light-font"
  >
    Welcome to your Favorite Pokemon Section! Discover, admire, and celebrate the Pokemon that hold a special place in your heart.
  </p>
  <div
    class="tools"
  >
    <div
      class="search_box light-input-bg"
    >
      <input
        class="light-font"
        type="text"
        value=""
      />
      <button
        disabled=""
      >
        <i
          class="fa-solid fa-spinner fa-spin"
        />
      </button>
    </div>
    <div
      class="select"
    >
      <select
        class="light-font light-input-bg"
        id="pokemon-select-filter"
        name="pokemon-select-filter"
      >
        <option
          value="all"
        >
          All
        </option>
        <option
          value="bug"
        >
          Bug
        </option>
        <option
          value="dark"
        >
          Dark
        </option>
        <option
          value="dragon"
        >
          Dragon
        </option>
        <option
          value="electric"
        >
          Electric
        </option>
        <option
          value="fairy"
        >
          Fairy
        </option>
        <option
          value="fighting"
        >
          Fighting
        </option>
        <option
          value="fire"
        >
          Fire
        </option>
        <option
          value="flying"
        >
          Flying
        </option>
        <option
          value="ghost"
        >
          Ghost
        </option>
        <option
          value="grass"
        >
          Grass
        </option>
        <option
          value="ground"
        >
          Ground
        </option>
        <option
          value="ice"
        >
          Ice
        </option>
        <option
          value="normal"
        >
          Normal
        </option>
        <option
          value="poison"
        >
          Poison
        </option>
        <option
          value="psychic"
        >
          Psychic
        </option>
        <option
          value="rock"
        >
          Rock
        </option>
        <option
          value="steel"
        >
          Steel
        </option>
        <option
          value="water"
        >
          Water
        </option>
      </select>
      <div>
        <i
          class="fa-solid fa-filter"
        />
      </div>
    </div>
    <div
      class="select"
    >
      <select
        class="light-font light-input-bg"
        id="pokemon-select-sort"
        name="pokemon-select-sort"
      >
        <option
          value="ascending"
        >
          Ascending
        </option>
        <option
          value="descending"
        >
          Descending
        </option>
      </select>
      <div>
        <i
          class="fa-solid fa-sort"
        />
      </div>
    </div>
  </div>
  <div
    class="pokemon_grid"
  >
    <article
      class="pokemon_card pokemon-type-grass"
    >
      <section>
        <div
          class="pokemon_header"
        >
          <h2>
            Bulbasaur
          </h2>
          <button>
            <i
              class="fa-regular fa-heart fa-xl"
            />
          </button>
        </div>
        <div
          class="pokemon_chips"
        >
          <div
            class="pokemon_chip"
          >
            Grass
          </div>
        </div>
        <div
          class="pokemon_bg_image_container"
        >
          <div
            class="pokeball_icon"
          >
            <img
              alt="pokeball icon"
              data-nimg="1"
              decoding="async"
              height="150"
              loading="lazy"
              src="/_next/image?url=%2Fimages%2Fpokeball-icon.png&w=384&q=75"
              srcset="/_next/image?url=%2Fimages%2Fpokeball-icon.png&w=256&q=75 1x, /_next/image?url=%2Fimages%2Fpokeball-icon.png&w=384&q=75 2x"
              style="color: transparent;"
              width="150"
            />
          </div>
          <div
            class="points_icon"
          >
            <img
              alt="points icon"
              data-nimg="1"
              decoding="async"
              height="50"
              loading="lazy"
              src="/_next/image?url=%2Fimages%2Fpoints-icon.png&w=128&q=75"
              srcset="/_next/image?url=%2Fimages%2Fpoints-icon.png&w=64&q=75 1x, /_next/image?url=%2Fimages%2Fpoints-icon.png&w=128&q=75 2x"
              style="color: transparent;"
              width="50"
            />
          </div>
          <div
            class="lines_icon"
          >
            <img
              alt="lines icon"
              data-nimg="1"
              decoding="async"
              height="50"
              loading="lazy"
              src="/_next/image?url=%2Fimages%2Flines-icon.png&w=128&q=75"
              srcset="/_next/image?url=%2Fimages%2Flines-icon.png&w=64&q=75 1x, /_next/image?url=%2Fimages%2Flines-icon.png&w=128&q=75 2x"
              style="color: transparent;"
              width="50"
            />
          </div>
          <div
            class="pokemon_image"
          >
            <img
              alt="pokemon image"
              data-nimg="1"
              decoding="async"
              fetchpriority="high"
              height="200"
              src="/_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2F1.png&w=640&q=75"
              srcset="/_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2F1.png&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2F1.png&w=640&q=75 2x"
              style="color: transparent;"
              width="200"
            />
          </div>
        </div>
      </section>
      <div
        class="pokemon_stats light-card-bg"
      >
        <p
          class="light-font"
        >
          Base Stats
        </p>
        <div
          class="base_stats"
        />
      </div>
    </article>
    <article
      class="pokemon_card pokemon-type-fire"
    >
      <section>
        <div
          class="pokemon_header"
        >
          <h2>
            Charizard
          </h2>
          <button>
            <i
              class="fa-regular fa-heart fa-xl"
            />
          </button>
        </div>
        <div
          class="pokemon_chips"
        >
          <div
            class="pokemon_chip"
          >
            Fire
          </div>
        </div>
        <div
          class="pokemon_bg_image_container"
        >
          <div
            class="pokeball_icon"
          >
            <img
              alt="pokeball icon"
              data-nimg="1"
              decoding="async"
              height="150"
              loading="lazy"
              src="/_next/image?url=%2Fimages%2Fpokeball-icon.png&w=384&q=75"
              srcset="/_next/image?url=%2Fimages%2Fpokeball-icon.png&w=256&q=75 1x, /_next/image?url=%2Fimages%2Fpokeball-icon.png&w=384&q=75 2x"
              style="color: transparent;"
              width="150"
            />
          </div>
          <div
            class="points_icon"
          >
            <img
              alt="points icon"
              data-nimg="1"
              decoding="async"
              height="50"
              loading="lazy"
              src="/_next/image?url=%2Fimages%2Fpoints-icon.png&w=128&q=75"
              srcset="/_next/image?url=%2Fimages%2Fpoints-icon.png&w=64&q=75 1x, /_next/image?url=%2Fimages%2Fpoints-icon.png&w=128&q=75 2x"
              style="color: transparent;"
              width="50"
            />
          </div>
          <div
            class="lines_icon"
          >
            <img
              alt="lines icon"
              data-nimg="1"
              decoding="async"
              height="50"
              loading="lazy"
              src="/_next/image?url=%2Fimages%2Flines-icon.png&w=128&q=75"
              srcset="/_next/image?url=%2Fimages%2Flines-icon.png&w=64&q=75 1x, /_next/image?url=%2Fimages%2Flines-icon.png&w=128&q=75 2x"
              style="color: transparent;"
              width="50"
            />
          </div>
          <div
            class="pokemon_image"
          >
            <img
              alt="pokemon image"
              data-nimg="1"
              decoding="async"
              fetchpriority="high"
              height="200"
              src="/_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2F6.png&w=640&q=75"
              srcset="/_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2F6.png&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2F6.png&w=640&q=75 2x"
              style="color: transparent;"
              width="200"
            />
          </div>
        </div>
      </section>
      <div
        class="pokemon_stats light-card-bg"
      >
        <p
          class="light-font"
        >
          Base Stats
        </p>
        <div
          class="base_stats"
        />
      </div>
    </article>
    <article
      class="pokemon_card pokemon-type-electric"
    >
      <section>
        <div
          class="pokemon_header"
        >
          <h2>
            Pikachu
          </h2>
          <button>
            <i
              class="fa-regular fa-heart fa-xl"
            />
          </button>
        </div>
        <div
          class="pokemon_chips"
        >
          <div
            class="pokemon_chip"
          >
            Electric
          </div>
        </div>
        <div
          class="pokemon_bg_image_container"
        >
          <div
            class="pokeball_icon"
          >
            <img
              alt="pokeball icon"
              data-nimg="1"
              decoding="async"
              height="150"
              loading="lazy"
              src="/_next/image?url=%2Fimages%2Fpokeball-icon.png&w=384&q=75"
              srcset="/_next/image?url=%2Fimages%2Fpokeball-icon.png&w=256&q=75 1x, /_next/image?url=%2Fimages%2Fpokeball-icon.png&w=384&q=75 2x"
              style="color: transparent;"
              width="150"
            />
          </div>
          <div
            class="points_icon"
          >
            <img
              alt="points icon"
              data-nimg="1"
              decoding="async"
              height="50"
              loading="lazy"
              src="/_next/image?url=%2Fimages%2Fpoints-icon.png&w=128&q=75"
              srcset="/_next/image?url=%2Fimages%2Fpoints-icon.png&w=64&q=75 1x, /_next/image?url=%2Fimages%2Fpoints-icon.png&w=128&q=75 2x"
              style="color: transparent;"
              width="50"
            />
          </div>
          <div
            class="lines_icon"
          >
            <img
              alt="lines icon"
              data-nimg="1"
              decoding="async"
              height="50"
              loading="lazy"
              src="/_next/image?url=%2Fimages%2Flines-icon.png&w=128&q=75"
              srcset="/_next/image?url=%2Fimages%2Flines-icon.png&w=64&q=75 1x, /_next/image?url=%2Fimages%2Flines-icon.png&w=128&q=75 2x"
              style="color: transparent;"
              width="50"
            />
          </div>
          <div
            class="pokemon_image"
          >
            <img
              alt="pokemon image"
              data-nimg="1"
              decoding="async"
              fetchpriority="high"
              height="200"
              src="/_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2F25.png&w=640&q=75"
              srcset="/_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2F25.png&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2F25.png&w=640&q=75 2x"
              style="color: transparent;"
              width="200"
            />
          </div>
        </div>
      </section>
      <div
        class="pokemon_stats light-card-bg"
      >
        <p
          class="light-font"
        >
          Base Stats
        </p>
        <div
          class="base_stats"
        />
      </div>
    </article>
  </div>
</div>
`;
